File: firstimagedemo/src/main/java/com/melayer/firstimagedemo/MainActivity.java
package com.melayer.firstimagedemo;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ProgressBar;
import com.melayer.firstimagedemo.adapter.RecyclerAdapter;
import com.melayer.firstimagedemo.models.NicePlaces;
import com.melayer.firstimagedemo.viewModels.MainActivityViewModel;

import java.util.List;

public class MainActivity extends AppCompatActivity {

    private FloatingActionButton fab;
    private RecyclerView recyclerView;
    private RecyclerAdapter recyclerAdapter;
    private ProgressBar progressBar;
    private MainActivityViewModel mainActivityViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        fab = findViewById(R.id.fab);
        recyclerView = findViewById(R.id.recyclerView);
        progressBar = findViewById(R.id.progressBar);

        mainActivityViewModel = ViewModelProviders.of(this).get(MainActivityViewModel.class);
        mainActivityViewModel.init();

        mainActivityViewModel.getNicePlaces().observe(this, new Observer<List<NicePlaces>>() {
            @Override
            public void onChanged(@Nullable List<NicePlaces> nicePlaces) {
                recyclerAdapter.notifyDataSetChanged();
            }
        });

        mainActivityViewModel.getIsUpdating().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(@Nullable Boolean aBoolean) {
                if (aBoolean){
                    showProgressBar();
                }
                else {
                    hideProgressBar();
                    recyclerView.smoothScrollToPosition(mainActivityViewModel.getNicePlaces().getValue().size());
                }
            }
        });

        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mainActivityViewModel.addNewValue(
                        new NicePlaces("Washington","https://i.imgur.com/ZcLLrkY.jpg")
                );
            }
        });

        initRecyclerView();
    }

    private void initRecyclerView(){
        recyclerAdapter = new RecyclerAdapter(this, mainActivityViewModel.getNicePlaces().getValue());
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(recyclerAdapter);
    }

    private void showProgressBar(){
        progressBar.setVisibility(View.VISIBLE);
    }

    private void hideProgressBar(){
        progressBar.setVisibility(View.INVISIBLE);
    }
}

File: app/src/main/java/com/melayer/myapplication/MainActivity.kt
package com.melayer.myapplication

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import android.widget.ProgressBar
import com.melayer.myapplication.adapter.RecyclerAdapter
import com.melayer.myapplication.models.NicePlaces
import com.melayer.myapplication.viewModels.MainActivityViewModel
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private var recyclerAdapter: RecyclerAdapter? = null
    private var progressBar: ProgressBar? = null
    private var mainActivityViewModel: MainActivityViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        mainActivityViewModel =
            ViewModelProviders.of(this).get<MainActivityViewModel>(MainActivityViewModel::class.java!!)
        mainActivityViewModel?.init()

        mainActivityViewModel?.nicePlaces?.observe(this,
            Observer<ArrayList<NicePlaces>> { recyclerAdapter?.notifyDataSetChanged() })

        mainActivityViewModel?.isUpdating?.observe(this, Observer<Boolean> { aBoolean ->
            if (aBoolean!!) {
                showProgressBar()
            } else {
                hideProgressBar()
                recyclerView?.smoothScrollToPosition(mainActivityViewModel?.nicePlaces?.value!!.size)
            }
        })

        fab?.setOnClickListener {
            mainActivityViewModel?.addNewValue(
                NicePlaces("Washington", "https://i.imgur.com/ZcLLrkY.jpg")
            )
        }
        initRecyclerView()
    }

    private fun initRecyclerView() {
        recyclerAdapter = RecyclerAdapter(this, mainActivityViewModel?.nicePlaces?.value!!)
        val layoutManager = LinearLayoutManager(this)
        recyclerView?.layoutManager = layoutManager
        recyclerView?.adapter = recyclerAdapter
    }

    private fun showProgressBar() {
        progressBar?.visibility = View.VISIBLE
    }

    private fun hideProgressBar() {
        progressBar?.visibility = View.INVISIBLE
    }
}